crt/fork.c

```diff
@@ -0,0 +1,569 @@
+**
+** _create_child_pthread()
+**
+** Create a new thread context from the current thread and return a pointer to
+** the new pthread structure. The pthread has the following layout.
+**
+**     [guard|stack|tls|tsd]
+**
+** Or:
+**     [ guard ]                td->guard_size | __default_guardsize
+**     [ stack ]                td->stack_size | __default_stacksize
+**     [ tls area | tcb ]       __libc.tls_size
+**     [ tsd area ]             __pthread_tsd_size
+**
+**==============================================================================
+*/
+
+struct pthread* _create_child_pthread()
+{
+    size_t size;
+    size_t size_rounded;
+    struct pthread* self = __pthread_self();
+    struct pthread* new;
+    uint8_t* map;
+    uint8_t* tsd;
+    uint8_t* tls;
+    uint8_t* stack;            /* bottom */
+    uint8_t* stack_limit;      /* top */
+    uint8_t* self_stack_limit; /* top */
+    size_t stack_size;
+    size_t guard_size;
+    void* process_stack = NULL;
+
+    _get_parent_stack((void**)&self_stack_limit, &stack_size);
+
+    guard_size = self->guard_size ? self->guard_size : __default_guardsize;
+    size = guard_size + stack_size + __libc.tls_size + __pthread_tsd_size;
+
+    myst_round_up(size, PAGE_SIZE, &size_rounded);
+
+    if (!(map = mmap(
+              NULL,
+              size_rounded,
+              PROT_READ | PROT_WRITE,
+              MAP_ANONYMOUS,
+              -1,
+              0)))
+        return NULL;
+
+    /* [guard|stack|tls|tsd] */
+    tsd = map + size - __pthread_tsd_size;
+    tls = tsd - __libc.tls_size;
+    stack = tsd - __libc.tls_size;
+    stack_limit = stack - stack_size;
+
+    new = __copy_tls(tls);
+    new->self = new;
+    new->map_base = map;
+    new->map_size = size;
+    new->stack = stack;
+    new->stack_size = stack - stack_limit;
+    new->guard_size = guard_size;
+    new->self = new;
+    new->tsd = (void*)tsd;
+
+    new->detach_state = DT_DETACHED;
+    new->robust_list.head = &new->robust_list.head;
+    new->canary = self->canary;
+    new->sysinfo = self->sysinfo;
+
+    /* copy over the stack if any */
+    memcpy(stack_limit, self_stack_limit, stack_size);
+
+    return new;
+}
+
+struct thread_args
+{
+    myst_jmp_buf_t env;
+    void* child_sp;
+    void* child_bp;
+    volatile pid_t pid;
+    struct pthread* child_pthread;
+
+    // [0] - pthread memory and stack
+    // [1] - this thread structure
+    struct mmap_info
+    {
+        void* mmap_ptr;
+        size_t mmap_ptr_size;
+    } unmap_on_exit[2];
+};
+
+/* internal musl function */
+extern int __clone(int (*func)(void*), void* stack, int flags, void* arg, ...);
+
+/* ATTN: arrange for this function to be called on exit */
+__attribute__((__unused__)) static void _thread_args_free(void* arg)
+{
+    struct thread_args* args = (struct thread_args*)arg;
+    /* ATTN: release child stack parent is not process thread */
+    free(args);
+}
+
+static int _valid_stack_address(const void* sp, const void* p)
```

Comment:
Seems this function is replaced by "_within".

Reply:
removed